File: frontend/mobile/lib/service/add_activity_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/api_config.dart';
import 'storage_helper.dart';
import 'auth_service.dart';

// Service สำหรับบันทึกกิจกรรมออกกำลังกาย
class AddActivityService {
  static String get baseUrl => ApiConfig.baseUrl;

  // บันทึกกิจกรรมออกกำลังกาย
  static Future<Map<String, dynamic>> logActivity({
    required String sportName,
    required int time,
  }) async {
    try {
      String? accessToken = await StorageHelper.getAccessToken();
      String? userId = await StorageHelper.getUserId();

      if (accessToken == null || accessToken.isEmpty) {
        throw Exception('No access token found. Please login again.');
      }

      if (userId == null || userId.isEmpty) {
        throw Exception('User ID not found. Please login again.');
      }

      final url = Uri.parse('$baseUrl/api/activity/$userId');

      var response = await http.post(
        url,
        headers: {
          'Authorization': 'Bearer $accessToken',
          'Content-Type': 'application/json',
        },
        body: jsonEncode({
          'sport_name': sportName,
          'time': time,
        }),
      );

      // ถ้า token หมดอายุ ให้ refresh แล้วลองใหม่
      if (response.statusCode == 401 || response.statusCode == 403) {
        try {
          accessToken = await AuthService.refreshAccessToken();

          response = await http.post(
            url,
            headers: {
              'Authorization': 'Bearer $accessToken',
              'Content-Type': 'application/json',
            },
            body: jsonEncode({
              'sport_name': sportName,
              'time': time,
            }),
          );
        } catch (e) {
          throw Exception('Session expired. Please login again.');
        }
      }

      if (response.statusCode == 201 || response.statusCode == 200) {
        final data = jsonDecode(response.body);

        return {
          'success': true,
          'message': data['message'] ?? 'Activity logged successfully',
          'sport_name': data['data']?['sport_name'] ?? sportName,
          'time': data['data']?['time'] ?? time,
          'calories_burned': data['data']?['calories_burned'] ?? 0,
          'total_burned': data['data']?['total_burned'] ?? 0,
        };
      } else if (response.statusCode == 404) {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Sport not found');
      } else if (response.statusCode == 400) {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Invalid input');
      } else {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Failed to log activity');
      }
    } catch (e) {
      if (e is Exception) rethrow;
      throw Exception('เกิดข้อผิดพลาด: ${e.toString()}');
    }
  }

  // ดึงรายการกีฬาทั้งหมด
  static Future<List<Map<String, dynamic>>> getSportsList() async {
    try {
      String? accessToken = await StorageHelper.getAccessToken();

      if (accessToken == null || accessToken.isEmpty) {
        throw Exception('No access token found. Please login again.');
      }

      final url = Uri.parse('$baseUrl/api/sports');

      var response = await http.get(
        url,
        headers: {
          'Authorization': 'Bearer $accessToken',
          'Content-Type': 'application/json',
        },
      );

      // ถ้า token หมดอายุ ให้ refresh แล้วลองใหม่
      if (response.statusCode == 401 || response.statusCode == 403) {
        accessToken = await AuthService.refreshAccessToken();

        response = await http.get(
          url,
          headers: {
            'Authorization': 'Bearer $accessToken',
            'Content-Type': 'application/json',
          },
        );
      }

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);

        if (data is List) {
          return data.map((sport) => sport as Map<String, dynamic>).toList();
        } else if (data['sports'] != null) {
          return (data['sports'] as List)
              .map((sport) => sport as Map<String, dynamic>)
              .toList();
        } else {
          throw Exception('Unexpected response format');
        }
      } else {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Failed to fetch sports list');
      }
    } catch (e) {
      if (e is Exception) rethrow;
      throw Exception('เกิดข้อผิดพลาด: ${e.toString()}');
    }
  }
}

File: frontend/web/src/services/add_activity_service.ts
// src/app/services/add_activity_service.ts
import api from './auth_service';
import { getNodeApiUrl } from '@/config/api.config';

const API_BASE_URL = getNodeApiUrl();

interface LogActivityResponse {
  success: boolean;
  message: string;
  sport_name: string;
  time: number;
  calories_burned: number;
  total_burned: number;
}

interface Sport {
  sport_id: number;
  sport_name: string;
  burn_out: number;
}

// ========================================
// Helper Functions
// ========================================
function getUserId(): number | null {
  if (typeof window === 'undefined') return null;

  const userStr = localStorage.getItem('user');
  if (!userStr) return null;

  try {
    const user = JSON.parse(userStr);
    return user.id || null;
  } catch {
    return null;
  }
}

// ========================================
// Activity Service
// ========================================
export const AddActivityService = {
  /**
   * Log an activity/sport session
   */
  async logActivity(sportName: string, time: number): Promise<LogActivityResponse> {
    try {
      const userId = getUserId();

      if (!userId) {
        throw new Error('User ID not found. Please login again.');
      }

      console.log('🏃 Calling activity API:', `${API_BASE_URL}/api/activity/${userId}`);
      console.log('📦 Request body:', { sport_name: sportName, time });

      // ใช้ axios instance จาก auth_service ที่มี interceptor อยู่แล้ว
      const response = await api.post(`/api/activity/${userId}`, {
        sport_name: sportName,
        time: time,
      });

      console.log('📡 Response data:', response.data);

      // แปลง response ให้ตรงกับ interface
      return {
        success: true,
        message: response.data.message || 'Activity logged successfully',
        sport_name: response.data.data?.sport_name || sportName,
        time: response.data.data?.time || time,
        calories_burned: parseFloat(response.data.data?.calories_burned) || 0,
        total_burned: parseFloat(response.data.data?.total_burned) || 0,
      };
    } catch (error: any) {
      console.error('❌ Error logging activity:', error);

      const errorMessage = error.response?.data?.message || error.message || 'Failed to log activity';
      throw new Error(errorMessage);
    }
  },

  /**
   * Get list of all sports
   */
  async getSportsList(): Promise<Sport[]> {
    try {
      const response = await api.get('/api/sports');

      if (Array.isArray(response.data)) {
        return response.data;
      } else if (response.data.sports && Array.isArray(response.data.sports)) {
        return response.data.sports;
      } else {
        throw new Error('Unexpected response format');
      }
    } catch (error: any) {
      console.error('❌ Error fetching sports list:', error);
      throw new Error(error.response?.data?.message || 'Failed to fetch sports list');
    }
  },
};
